iina/JavasctiptDevTool.swift

```diff
@@ -0,0 +1,506 @@
+//
+//  JavasctiptDevTool.swift
+//  iina
+//
+//  Created by Hechen Li on 11/30/22.
+//  Copyright © 2022 lhc. All rights reserved.
+//
+
+import SwiftUI
+import JavaScriptCore
+import WebKit
+
+
+// MARK: - Menu Controller
+
+@available(macOS 12.0, *)
+extension MenuController {
+  func createMenuItem(fromPluginInstance inst: JavascriptPluginInstance, tag: Int) -> NSMenuItem {
+    let item = NSMenuItem()
+    if inst.isGlobal {
+      item.title = "\(inst.plugin.name) (Global)"
+    } else {
+      item.title = inst.plugin.name
+    }
+    item.representedObject = inst
+    item.tag = JavasctiptDevTool.JSMenuItemInstance
+    item.target = self
+    item.action = #selector(openJavascriptDevTool)
+    return item
+  }
+
+  @objc func openJavascriptDevTool(_ sender: NSMenuItem) {
+    switch (sender.tag) {
+    case JavasctiptDevTool.JSMenuItemInstance:
+      if let inst = sender.representedObject as? JavascriptPluginInstance {
+        createJavascriptDevToolWindow(forInstance: inst, title: sender.title)
+      }
+      //    case JSMenuItemWebView:
+      //      is let webView = sender.representedObject as? WKWebView {
+      //        createJavascriptDevToolWindow(forWebView: webView)
+      //      }
+    default:
+      break
+    }
+  }
+}
+
+
+// MARK: - Data
+
+fileprivate struct JSEvent: Identifiable {
+  enum Result {
+    case nothing
+    case number(NSNumber)
+    case string(String)
+    case boolean(Bool)
+    case array(String, [(String, String)])
+    case null
+    case undefined
+    case opaqueObject(String)
+    case object(String, [(String, String)])
+    case exception(String, String)
+    case log(String, Logger.Level)
+  }
+  let id: Int
+  let index: Int
+  let prompt: String?
```

Comment:
I am not sure I entirely like the programming model where you mix prompts with results. Can these be two different types?

Reply:
I found no elegant way to store mixed `Identifiable` objects. Do you have better suggestions?

Reply:
Does `[Identifiable]` work?